### PR TITLE
Feature/add topics tools 3

### PR DESCRIPTION
**Pull request by redhug1**
Adjust dp-data-tools/topics-tools/gen-topics-database to write highlighted links on a topics page into spotlight on corresponding content page in content collection and create content link in the topics document in topics collection.

### CODE REVIEW

File: topics-tools/gen-topics-database/main.go [hunk 1]
```diff
@@ -1339,7 +1345,7 @@ func getPage(parentID int, graphVizFile io.Writer, bodyTextFile io.Writer, check
 					gvLink = strings.ReplaceAll(gvLink, "/", "_")
 					gvLinkLabel := "/" + strings.ReplaceAll(gvLink, "_", "\\n/")
 					if link.valid {
-						_, err = fmt.Fprintf(graphVizFile, "    subgraph %s {\n        %s [shape = box, color=\"#30A8A0\", style=filled, label = \"%s\"]\n    }\n", gvLink, gvLink, fmt.Sprintf("Topic - Highlighted:\\n")+gvLinkLabel)
+						_, err = fmt.Fprintf(graphVizFile, "    subgraph %s {\n        %s [shape = box, color=\"#30A8A0\", style=filled, label = \"%s\"]\n    }\n", gvLink, gvLink, fmt.Sprintf("Topic - Highlighted: %s\\n", link.linkType)+gvLinkLabel)
```
Comment (daiLlew) [inline topics-tools/gen-topics-database/main.go:1348]: Consider creating vars for these complex format strings.  

As someone reading this code for the first time I have no clue what this means:
`"    subgraph %s {\n        %s [shape = box, color=\"#30A8A0\", style=filled, label = \"%s\"]\n    }\n"

A well named variable will help to remove that ambiguity and keep the code short/readable.

File: topics-tools/gen-topics-database/main.go [hunk 2]
```diff
@@ -4422,10 +4454,20 @@ func populateTopicAndContentStructs(topics []TopicResponseStore, content []Conte
...
+						}
+					}
+
+					if pageType[id] != pageTopicAndContent {
+						if pageType[id] == pageTopic {
+							// add content to topic
+							pageType[id] = pageTopicAndContent
+						} else {
+							pageType[id] = pageContent
+						}
```
Comment (daiLlew) [inline topics-tools/gen-topics-database/main.go:4468]: These appear to be duplicates of code written previously (line 4381 & 4354). Is this something that can be abstracted out into a function and reused?

File: topics-tools/gen-topics-database/main.go [hunk 3]
```diff
@@ -4242,20 +4248,34 @@ func getTerminationNodeData(data *DataResponse, parentTopicNumber int, parentFul
 	return info
 }
 
-func getNodeData(data *DataResponse, parentTopicNumber int, parentFullURI string, bodyTextFile io.Writer, checkFile io.Writer) {
+func getNodeData(data *DataResponse, parentTopicNumber int, parentFullURI string, bodyTextFile io.Writer, checkFile io.Writer) bool {
+	var anyValid bool
```
Comment (daiLlew) [inline topics-tools/gen-topics-database/main.go:4252]: Consider explicitly setting this to false for clarity
```
anyValid := false
```
It makes it immediately clear to anyone reading this what the default/initial value of this var is. The zero value of a bool in Go is `false` but I wasn't 100% sure and had to look it up to make sure my understanding of this code was correct.

**Review by daiLlew — COMMENTED** [review]
I've added some points to consider. The other comment I have is consider breaking up some of these functions as some of them are very very long. This is usually a sign that they are doing more than 1 task which is not optimal. 

There is no hard rule of function size and there is an element of personal opinion _but_ consider if there is requirement for it to be 1 large function or if the same can be achieved by  making each steps within the function a function of its own? 

In my opinion (feel free to disagree). Smaller functions are easier to read as their intent is much easier to understand. Also the make function read like a set of instructions `doX, then doY etc etc`.

It's not always possible and doesn't always make sense but at least if you've considered it you can justify it. Breaking a large task into smaller sections also helps to reduce code duplication.

File: topics-tools/gen-topics-database/main.go [hunk 4]
```diff
@@ -1325,7 +1326,12 @@ func getPage(parentID int, graphVizFile io.Writer, bodyTextFile io.Writer, check
 		check(err)
 
 		// ===========================================
-		getNodeData(&data, returnedIndexNumber, fullURI, bodyTextFile, checkFile)
+		if getNodeData(&data, returnedIndexNumber, fullURI, bodyTextFile, checkFile) {
+			// This page has 'highlighted' type which has been re-maped to spotlight ... that is on a content page
+			// so, Tag page as also having content:
+			listOfDuplicateInfo = append(listOfDuplicateInfo, duplicateInfo{id: returnedIndexNumber, pageType: pageContent, parentURI: parentURI, shortURI: shortURI})
+			listOfPageData = append(listOfPageData, pageData{id: returnedIndexNumber, subSectionIndex: parentID, pageType: pageContent, uriStatus: pageContent, shortURI: shortURI, parentURI: parentURI, fixedPayload: fixedJSON})
```
[the remark this answers is not in the document]
Comment (redhug1) [reply topics-tools/gen-topics-database/main.go:1333]: Good point, adjustments made.

File: topics-tools/gen-topics-database/main.go [hunk 5]
```diff
@@ -1325,7 +1326,12 @@ func getPage(parentID int, graphVizFile io.Writer, bodyTextFile io.Writer, check
 		check(err)
 
 		// ===========================================
-		getNodeData(&data, returnedIndexNumber, fullURI, bodyTextFile, checkFile)
+		if getNodeData(&data, returnedIndexNumber, fullURI, bodyTextFile, checkFile) {
+			// This page has 'highlighted' type which has been re-maped to spotlight ... that is on a content page
+			// so, Tag page as also having content:
+			listOfDuplicateInfo = append(listOfDuplicateInfo, duplicateInfo{id: returnedIndexNumber, pageType: pageContent, parentURI: parentURI, shortURI: shortURI})
```
[the remark this answers is not in the document]
Comment (redhug1) [reply topics-tools/gen-topics-database/main.go:1332]: done.

File: topics-tools/gen-topics-database/main.go (hunk 1 above)
Comment (redhug1) [reply topics-tools/gen-topics-database/main.go:1348]: This code is building up a graphviz doc on the fly and putting those strings into vars would make them even harder to understand as you need to see the whole line in context ... honest - this is for what its doing time optimal.

File: topics-tools/gen-topics-database/main.go (hunk 2 above)
Comment (redhug1) [reply topics-tools/gen-topics-database/main.go:4468]: No, the inner part of the iff is different, they are ok.

File: topics-tools/gen-topics-database/main.go (hunk 3 above)
Comment (redhug1) [reply topics-tools/gen-topics-database/main.go:4252]: I can see your point, but others would then pick me up saying to do it the way it currently is, so this one will be left as is.

Comment (eldeal) [reply topics-tools/gen-topics-database/main.go:4360]: I would agree with dave, setting flags to their default state explicitly is much clearer

**Comment by eldeal** [comment]
> this function processes ~27 different page types and a number of them have issues and specific code to cope

To me that is exactly where I'd look first to factor code out. If the inner workings of each case in a switch is long, those are also places I'd look to factor out. Our coding standards prefer the separation of functions into reasonable units (benefits: easier to test, understand, more clearly named) and we're willing to invest the time to ensure our code meets the standards. 

If the task has taken longer than expected, then this is something we can feedback through retro and planning to learn from for future definitely.

**Review by gedge — APPROVED** [review]

**Pull request merged**